### PR TITLE
sahara: use barbican key manager made configurable

### PR DESCRIPTION
**Pull request by matelakat**
People can configure on the barclamp's UI if they want to use Barbican's
key manager.

### CODE REVIEW

**Comment by matelakat** [comment]
[P4: Add Barbican support to sahara barclamp](https://trello.com/c/5vY3AtXG/121-p4-add-barbican-support-to-sahara-barclamp)

**Comment by matelakat** [comment]
When enabling barbican, I get this:
http://pastebin.suse.de/17196

**Comment by matelakat** [comment]
wait until https://bugs.launchpad.net/sahara/+bug/1641897 is merged.

**Comment by toabctl** [comment]
> wait until https://bugs.launchpad.net/sahara/+bug/1641897 is merged.

Where is the review for that? I can  not find it in the bug. And can we just add the patch to our package for now?

**Comment by nicolasbock** [comment]
@toabctl it should be this one:

https://review.openstack.org/#/c/397681/

**Comment by matelakat** [comment]
@toabctl @nicolasbock - upstream change hit trunk - Let's merge this change!

**Comment by matelakat** [comment]
@toabctl - If I recall correclty, the policy is that we need 2 reviews for a change to be merged - could you take a look at it?

**Comment by toabctl** [comment]
> Now, no idea if we can check if barbican is deployed before showing the field or it needs to be implemented in the cookbook directly, which may be confusing to customers.

It can be done in the crowbar model . there are validations for other things. Just look there.

**Comment by vuntz** [comment]
Marked as WIP as there's still some work to do here.

**Comment by Itxaka** [comment]
rebased!

**Comment by Itxaka** [comment]
Added constraint for barbicand to be deployed before aloowing barbican to be set as key manager

File: crowbar_framework/app/models/sahara_service.rb [hunk 1]
```diff
@@ -84,6 +84,14 @@ def create_proposal
 
   def validate_proposal_after_save(proposal)
     validate_one_for_role proposal, "sahara-server"
+
+    # check if barbican is deployed in case we are using use_barbican_key_manager
+    if proposal["attributes"]["sahara"]["use_barbican_key_manager"]
+      if RoleObject.active("barbican").empty?
```
Comment (toabctl) [inline crowbar_framework/app/models/sahara_service.rb]: Are you sure that this works? magnum is doing it differently: https://github.com/crowbar/crowbar-openstack/blob/master/crowbar_framework/app/models/magnum_service.rb#L98 . Would be nice to use the same style here.

Comment (Itxaka) [inline crowbar_framework/app/models/sahara_service.rb]: I can confirm that it works as expected, but I'll change it to be consistent across barclamps.

File: crowbar_framework/app/models/sahara_service.rb [hunk 2]
```diff
@@ -84,6 +84,14 @@ def create_proposal
 
   def validate_proposal_after_save(proposal)
     validate_one_for_role proposal, "sahara-server"
+
+    # check if barbican is deployed in case we are using use_barbican_key_manager
+    if proposal["attributes"]["sahara"]["use_barbican_key_manager"]
```
Comment (sayalilunkad) [inline crowbar_framework/app/models/sahara_service.rb]: Can use this instead: proposal["attributes"][@bc_name]["use_barbican_key_manager"] as is in the other barclamps

File: crowbar_framework/app/models/sahara_service.rb [hunk 3]
```diff
@@ -84,6 +84,14 @@ def create_proposal
...
+      if NodeObject.find("roles:barbican-controller").empty?
+        validation_error I18n.t("barclamp.sahara.validation.barbican_deployed")
```
Comment (sayalilunkad) [inline crowbar_framework/app/models/sahara_service.rb]: Same here too.

File: crowbar_framework/app/models/sahara_service.rb (hunk 2 above)
Comment (Itxaka) [inline crowbar_framework/app/models/sahara_service.rb]: done!

File: crowbar_framework/app/models/sahara_service.rb (hunk 3 above)
Comment (Itxaka) [inline crowbar_framework/app/models/sahara_service.rb]: done

**Comment by Itxaka** [comment]
@toabctl @sayalilunkad @nicolasbock ready for re-review, all comments addressed!

File: chef/data_bags/crowbar/migrate/sahara/102_add_use_barbican_key_manager.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+def upgrade(ta, td, a, d)
+  a["use_barbican_key_manager"] = ta["use_barbican_key_manager"]
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/sahara/102_add_use_barbican_key_manager.rb]: Now that we need a backport of this, you need to check if the key already exist before setting it to the default

Comment (Itxaka) [inline chef/data_bags/crowbar/migrate/sahara/102_add_use_barbican_key_manager.rb]: Ah, good point!

**Comment by toabctl** [comment]
Backport is https://github.com/crowbar/crowbar-openstack/pull/1053

**Pull request merged**